README.md
![LN Invoice Logo](./images/ln-invoice-transparent.png)

Parse Bitcoin Lightning Network invoices (payment requests) in Kotlin. This library implements decoding of invoices as per
[BOLT-11: # Invoice Protocol for Lightning Payments](https://github.com/lightning/bolts/blob/master/11-payment-encoding.md#bolt-11-invoice-protocol-for-lightning-payments)

[<img src="https://img.shields.io/nexus/r/app.cash.lninvoice/ln-invoice.svg?label=latest%20release&server=https%3A%2F%2Foss.sonatype.org"/>](https://central.sonatype.com/namespace/app.cash.lninvoice)

## Getting Started

On the [Sontaype page for ln-invoice](https://central.sonatype.com/namespace/app.cash.lninvoice), choose the latest version
of `ln-invoice` and follow the instructions for inclusion in your build tool.

### Parsing a payment request

```kotlin
const val sample = "lnbc25m1pvjluezpp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqdq" +
  "5vdhkven9v5sxyetpdeessp5zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zygs9q5sqqqqqq" +
  "qqqqqqqqqpqsq67gye39hfg3zd8rgc80k32tvy9xk2xunwm5lzexnvpx6fd77en8qaq424dxgt56cag2dpt359" +
  "k3ssyhetktkpqh24jqnjyw6uqd08sgptq44qu"

val request: ErrorOr<PaymentRequest> = PaymentRequest.parse(sample)
```

See `app.cash.lninvoice.PaymentRequestTest` for more parsing examples, lifted directly from the BOLT-11 spec.

LN-Invoice is function programming safe by default, using [Quiver](https://github.com/cashapp/quiver)
(an extension of [Arrow](https://arrow-kt.io/)). If you prefer to deal with thrown exceptions when parsing,
simply add `.orThrow()`.

```kotlin
val request: PaymentRequest = PaymentRequest.parse(sample).orThrow()
```

Fields in `PaymentRequest` instances that contain binary data use 
[Okio's ByteString](https://square.github.io/okio/).
If you need this as a byte array, call `.toByteArray()`

```kotlin
val signature: ByteString = invoice.signature
val sigBytes: ByteArray = signature.toByteArray()
```


## Documentation

The [API documentation](https://cashapp.github.io/ln-invoice) is published with each release.

## Changelog

See a list of changes in each release in the [CHANGELOG](CHANGELOG.md).

## Contributing

### Building

1. Install [Hermit](https://cashapp.github.io/hermit/), the hermetic project tooling tool.

2. Use `gradle` to run all tests locally:

```shell
gradle build
```

For more details on contributing, see the [CONTRIBUTING](CONTRIBUTING.md) guide.


CHANGELOG.md
Change Log
==========

Next release
----------------------------


Version 0.0.1 *(2023-07-25)*
----------------------------

* Initial public release from internal snapshot. Parse Bech32 data & lightning invoices. <Jem Mawson> 


lib/src/test/kotlin/app/cash/lninvoice/Invoices.kt
/*
 * Copyright (c) 2023 Block, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.lninvoice

import app.cash.quiver.extensions.orThrow

object Invoices {

  const val sample = "lnbc25m1pvjluezpp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqdq" +
    "5vdhkven9v5sxyetpdeessp5zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zygs9q5sqqqqqq" +
    "qqqqqqqqqpqsq67gye39hfg3zd8rgc80k32tvy9xk2xunwm5lzexnvpx6fd77en8qaq424dxgt56cag2dpt359" +
    "k3ssyhetktkpqh24jqnjyw6uqd08sgptq44qu"

  const val sampleWithPaymentHash =
    "lnbc9678785340p1pwmna7lpp5gc3xfm08u9qy06djf8dfflhugl6p7lgza6dsjxq454gx" +
      "hj9t7a0sd8dgfkx7cmtwd68yetpd5s9xar0wfjn5gpc8qhrsdfq24f5ggrxdaezqsnvda3kkum5wfjkzmfqf3j" +
      "kgem9wgsyuctwdus9xgrcyqcjcgpzgfskx6eqf9hzqnteypzxz7fzypfhg6trddjhygrcyqezcgpzfysywmm5y" +
      "pxxjemgw3hxjmn8yptk7untd9hxwg3q2d6xjcmtv4ezq7pqxgsxzmnyyqcjqmt0wfjjq6t5v4khxsp5zyg3zyg" +
      "3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zygsxqyjw5qcqp2rzjq0gxwkzc8w6323m55m4jyxcjwmy" +
      "7stt9hwkwe2qxmy8zpsgg7jcuwz87fcqqeuqqqyqqqqlgqqqqn3qq9q9qrsgqrvgkpnmps664wgkp43l22qsgd" +
      "w4ve24aca4nymnxddlnp8vh9v2sdxlu5ywdxefsfvm0fq3sesf08uf6q9a2ke0hc9j6z6wlxg5z5kqpu2v9wz"

  const val pubkeyRecoveryTest =
    "lnbc69420n1psmnpx3pp5uaycaj6z4xta6f9235py6xzxekanr593hhycm4xz5uvfmdenps4qdpqwp6ky6m90ys" +
      "8yetrdamx2uneyp6x2um5cqzpgxqrrsssp5feksfv22atsf9ej9z6h6q4hg6uk7zyt454g7hhjt5ej8t5kw" +
      "qv6s9qyyssqy38g7quvc37egftsnw9urvztt7p5xsajhdrvxyseytv5feyw7fnkj599fr9k0tw3atql5p4g" +
      "9esuv69qakj3q85r7mwx4crsuwjmahgpxqnhrw"

  const val signatureOverflowTest = "lnbc1u1psa8wkepp5ly0f09hm3757w7pww6s78nykge4qea4kjas6vj" +
    "tfjuaaycucpyzqdqqcqzpgxqyz5vqsp5uxllhnws7kfr85exa7phjaadmgmtau8pqeffggqm6t46jxma987q9" +
    "qyyssqhhfttgqemkd2q5xhgwd3gdcqzqsmh5mlclkl3gjfy7lgygf7f5746cfwplpvhpzucpeg4ptzfk9k94c" +
    "6w8f4dagjkjhz03lklat542cpvm5ps5"

  const val sampleWithDescriptionHash = "lnbc20m1pvjluezsp5zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg" +
    "3zyg3zyg3zyg3zyg3zygspp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqhp58yjmda" +
    "n79s6qqdhdzgynm4zwqd5d7xmw5fk98klysy043l2ahrqs9qrsgq7ea976txfraylvgzuxs8kgcw23ezlrszf" +
    "nh8r6qtfpr6cxga50aj6txm9rxrydzd06dfeawfk6swupvz4erwnyutnjq7x39ymw6j38gp7ynn44"

  const val sampleWithDescriptionAndDescriptionHash = "lnbc20m1pvjluezpp5qqqsyqcyq5rqwzqfq" +
    "qqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqsp5zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg" +
    "3zygsdq823jhxaqhp58yjmdan79s6qqdhdzgynm4zwqd5d7xmw5fk98klysy043l2ahrqs9qrsgqxqrrsscqp" +
    "fr6twf5rlcx42v9x9pq3upsl0m24z2gtkwmkzcwehx2rcqrdgs5540lz8m0lhs3z7t8luva6c5hlg6l5jsw4x" +
    "3lnwuauqvxzmmcqen3gpj7tnq2"

  const val sampleWithUnknownTags = "lnbc25m1pvjluezpp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfq" +
    "qqsyqcyq5rqwzqfqypqdq5vdhkven9v5sxyetpdeessp5zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3" +
    "zyg3zyg3zygs9q5sqqqqqqqqqqqqqqqqsgq2qrqqqfppnqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqppnqqqqq" +
    "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqpp4qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq" +
    "qqqqqqqqqqqqqqqqqhpnqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqhp4qqqqqqqqqqq" +
    "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqspnqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq" +
    "qqqqqqqqqqqsp4qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqnp5qqqqqqqqqqqqqqq" +
    "qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqnpkqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq" +
    "qqqqqqqqqz599y53s3ujmcfjp5xrdap68qxymkqphwsexhmhr8wdz5usdzkzrse33chw6dlp3jhuhge9ley7j" +
    "2ayx36kawe7kmgg8sv5ugdyusdcqzn8z9x"

  val sampleDecoded: Bech32Data by lazy { sample.toBech32Data().orThrow() }
}


lib/src/test/kotlin/app/cash/lninvoice/BitReaderTest.kt
/*
 * Copyright (c) 2023 Block, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.lninvoice

import app.cash.lninvoice.Invoices.sampleWithDescriptionHash
import app.cash.quiver.extensions.orThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe

class BitReaderTest : StringSpec({

  val bolt11Sample =
    "lnbc1pvjluezsp5zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zyg3zygspp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqc" +
      "yq5rqwzqfqypqdpl2pkx2ctnv5sxxmmwwd5kgetjypeh2ursdae8g6twvus8g6rfwvs8qun0dfjkxaq9qrsgq357wnc5r2ueh7ck6q93d" +
      "j32dlqnls087fxdwk8qakdyafkq3yap9us6v52vjjsrvywa6rt52cm9r9zqt8r2t7mlcwspyetp5h2tztugp9lfyql"

  "testing bolt-11 example" {
    val data = bolt11Sample.toBech32Data().orThrow().payload
    val reader = BitReader(data.substring(0, data.size - 104))
    reader.long(7) shouldBe 1496314658L // timestamp
    reader.int(1) shouldBe 16 // 's' field
    reader.int(2) shouldBe 52 // data length
    reader.byteString(256).hex() shouldBe "1111111111111111111111111111111111111111111111111111111111111111"
    reader.int(1) shouldBe 1
    reader.int(2) shouldBe 52
    reader.byteString(256).hex() shouldBe "0001020304050607080900010203040506070809000102030405060708090102"
    reader.int(1) shouldBe 13
    reader.int(2) shouldBe 63
    reader.text(63) shouldBe "Please consider supporting this project"
    reader.int(1) shouldBe 5
    reader.int(2) shouldBe 3
    reader.bits(3) shouldBe setOf(8, 14)
  }

  "testing bolt-11 example with description hash" {
    val data = sampleWithDescriptionHash.toBech32Data().orThrow().payload
    val reader = BitReader(data.substring(0, data.size - 104))
    reader.long(7) shouldBe 1496314658 // timestamp
    reader.int(1) shouldBe 16 // 's' field
    reader.int(2) shouldBe 52 // data length
    reader.byteString(256).hex() shouldBe "1111111111111111111111111111111111111111111111111111111111111111"
    reader.int(1) shouldBe 1
    reader.int(2) shouldBe 52
    reader.byteString(256).hex() shouldBe "0001020304050607080900010203040506070809000102030405060708090102"
    reader.int(1) shouldBe 23 // 'h' field
    reader.int(2) shouldBe 52 // data length
    reader.byteString(256).hex() shouldBe "3925b6f67e2c340036ed12093dd44e0368df1b6ea26c53dbe4811f58fd5db8c1"
    reader.int(1) shouldBe 5
    reader.int(2) shouldBe 3
    reader.bits(3) shouldBe setOf(8, 14)
  }

  "testing uneven bytes with padding" {
    val data = "lnurl1xyerxdu27jy".toBech32Data().orThrow().payload
    val reader = BitReader(data)
    reader.byteString(24).string(Charsets.UTF_8) shouldBe "123"
  }
})


lib/src/test/kotlin/app/cash/lninvoice/Bech32DataTest.kt
/*
 * Copyright (c) 2023 Block, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.lninvoice

import app.cash.lninvoice.Bech32Data.Companion.Encoding
import io.kotest.assertions.arrow.core.shouldBeLeft
import io.kotest.assertions.arrow.core.shouldBeRight
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.should
import io.kotest.matchers.shouldBe
import io.kotest.property.Exhaustive
import io.kotest.property.checkAll
import io.kotest.property.exhaustive.collection
import java.util.Locale

class Bech32DataTest : StringSpec({

  "valid bech32 can serde" {
    checkAll(
      Exhaustive.collection(
        listOf(
          "A12UEL5L",
          "a12uel5l",
          "an83characterlonghumanreadablepartthatcontainsthenumber1andtheexcludedcharactersbio1tt5tgs",
          "abcdef1qpzry9x8gf2tvdw0s3jn54khce6mua7lmqqqxw",
          "11qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqc8247j",
          "split1checkupstagehandshakeupstreamerranterredcaperred2y9e3w",
          "?1ezyfcl"
        )
      )
    ) { encoded ->
      encoded.toBech32Data().shouldBeRight() should { data ->
        data.encoding shouldBe Encoding.BECH32
        data.encoded shouldBe encoded.lowercase(Locale.ROOT)
      }
    }
  }

  "valid bech32m can serde" {
    checkAll(
      Exhaustive.collection(
        listOf(
          "A1LQFN3A",
          "a1lqfn3a",
          "an83characterlonghumanreadablepartthatcontainsthetheexcludedcharactersbioandnumber11sg7hg6",
          "abcdef1l7aum6echk45nj3s0wdvt2fg8x9yrzpqzd3ryx",
          "11llllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllllludsr8",
          "split1checkupstagehandshakeupstreamerranterredcaperredlc445v",
          "?1v759aa"
        )
      )
    ) { encoded ->
      encoded.toBech32Data().shouldBeRight() should { data ->
        data.encoding shouldBe Encoding.BECH32M
        data.encoded shouldBe encoded.lowercase(Locale.ROOT)
      }
    }
  }

  "invalid bech32(m) cannot parse" {
    checkAll(
      Exhaustive.collection(
        listOf(
          " 1nwldj5", // HRP character out of range
          "${0x7f}1axkwrx", // HRP character out of range
          "${0x80}1eym55h", // HRP character out of range
          "an84characterslonghumanreadablepartthatcontainsthenumber1andtheexcludedcharactersbio1569pvx",
          "pzry9x0s0muk", // No separator character
          "1pzry9x0s0muk", // Empty HRP
          "x1b4n0q5v", // Invalid data character
          "li1dgmt3", // Too short checksum
          "de1lg7wt${0xff}", // Invalid character in checksum
          "A1G7SGD8", // checksum calculated with uppercase form of HRP
          "10a06t8", // empty HRP
          "1qzzfhee" // empty HRP
        )
      )
    ) { encoded ->
      encoded.toBech32Data().shouldBeLeft()
    }
  }
})


lib/src/main/kotlin/app/cash/lninvoice/Network.kt
/*
 * Copyright (c) 2023 Block, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.lninvoice

import app.cash.quiver.extensions.ErrorOr
import arrow.core.left
import arrow.core.right

/**
 * Represents allowed bitcoin networks as defined by SLIP-0173.
 */
enum class Network {
  MAIN,
  TEST;

  companion object {
    fun parse(value: String): ErrorOr<Network> = when (value) {
      "bc" -> MAIN.right()
      "tb" -> TEST.right()
      else -> UnknownNetworkException(value).left()
    }
  }
}

class UnknownNetworkException(network: String) : Throwable("$network is unknown and cannot be parsed")


lib/src/main/kotlin/app/cash/lninvoice/BitReader.kt
/*
 * Copyright (c) 2023 Block, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.lninvoice

import okio.ByteString
import okio.ByteString.Companion.toByteString
import java.math.BigInteger
import java.time.Instant

class BitReader(
  private val data: ByteString
) {

  private var byteIndex: Int = 0

  /** Read a timestamp by parsing `bytes` bytes into a long and considering it to be seconds since epoch */
  fun timestamp(bytes: Int): Instant = Instant.ofEpochSecond(long(bytes))

  /** Read a list of tagged fields until all data is consumed */
  fun taggedFields(): List<TaggedField> {
    tailrec fun keepOnGoing(acc: List<TaggedField>): List<TaggedField> =
      if (byteIndex >= data.size) acc else keepOnGoing(acc.plus(taggedField()))
    return keepOnGoing(emptyList())
  }

  /** Read the next tagged field */
  private fun taggedField(): TaggedField {
    val tag = int(1)
    val size = int(2)
    val bytes = bytes(size)
    return TaggedField(tag, size, bytes)
  }

  /** Read an integer by parsing `bytes` bytes, taking the lower 5 bits from each and concatenating them */
  internal fun int(bytes: Int): Int =
    bytes(bytes).toByteArray().fold(0) { acc, b ->
      (acc shl 5) + b
    }

  /** Read a long by parsing `bytes` bytes, taking the lower 5 bits from each and concatenating them */
  internal fun long(bytes: Int): Long =
    bytes(bytes).toByteArray().fold(0L) { acc, b ->
      (acc shl 5) + b
    }

  /** Capture the raw bytes */
  private fun bytes(bytes: Int): ByteString {
    val ans = data.substring(byteIndex, byteIndex + bytes)
    byteIndex += bytes
    return ans
  }

  /** Read the lower 5 bits from each byte until `bits` bits have been read, concatenate & return the resulting bytes */
  internal fun byteString(bits: Int): ByteString {
    tailrec fun bitList(remaining: Int, bits: String = ""): String =
      if (remaining == 0) {
        bits
      } else {
        val nextByte = data[byteIndex++].toInt()
        val bitsToRead = remaining.coerceAtMost(5)
        val bitString = Integer.toBinaryString(nextByte).padStart(5, '0').take(bitsToRead)
        bitList((remaining - 5).coerceAtLeast(0), bits + bitString)
      }
    return bitList(bits).chunked(8).map { Integer.parseInt(it, 2).toByte() }.toByteArray().toByteString()
  }

  /** Read `bytes` bytes (concatenating the lower 5 bits of each, as per `byteString`) and convert into UTF_8 text */
  internal fun text(bytes: Int): String {
    val bits = bytes * 5
    return byteString(bits - bits % 8).string(Charsets.UTF_8)
  }

  /** Parse a number of `bytes` bytes and returns a set of ints indicating which bits in that number are set */
  internal fun bits(bytes: Int): Set<Int> {
    val bi = BigInteger.valueOf(long(bytes))
    return (0..bi.bitLength()).fold(emptySet()) { acc, i ->
      if (bi.testBit(i)) acc + i else acc
    }
  }
}


lib/src/main/kotlin/app/cash/lninvoice/TaggedField.kt
/*
 * Copyright (c) 2023 Block, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.lninvoice

import okio.ByteString

data class TaggedField(val tag: Int, val size: Int, val data: ByteString)

enum class FieldTags(val tag: Int) {
  PAYMENT_HASH(1),
  EXTRA_ROUTING_INFO(3),
  FEATURE_BITS(5),
  EXPIRY(6),
  FALLBACK_ON_CHAIN_ADDRESS(9),
  DESCRIPTION(13),
  SECRET(16),
  PAYEE_NODE(19),
  DESCRIPTION_HASH(23),
  MIN_FINAL_CLTV_EXPIRY_DELTA(24),
  METADATA(27);

  companion object {
    fun validTags() = FieldTags.values().map { it.tag }
  }
}


lib/src/main/kotlin/app/cash/lninvoice/BitcoinAmount.kt
/*
 * Copyright (c) 2023 Block, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.lninvoice

import java.math.BigDecimal
import java.math.RoundingMode

/** A quantity of Bitcoin, internally denominated in satoshi, with a pico-btc remainder */
data class BitcoinAmount(
  /** The base unit of account in satoshis (10^-8 Bitcoin, or 100m satoshi per Bitcoin */
  val satoshi: Long,
  /** If necessary, any sub-satoshi remainder in pico btc (10^-12 Bitcoin, or 10k pico per satoshi) */
  val picoRemainder: Int = 0
) {

  /** Provide the amount denominated in BTC */
  fun bitcoin(): BigDecimal =
    BigDecimal(satoshi).divide(SATS_PER_BTC).plus(remainderAsBigDecimal).stripTrailingZeros()

  /** Provide the amount denominated in pico btc (10,000th of a satoshi) */
  fun pico(): Long = BigDecimal(satoshi)
    .multiply(PICO_PER_SAT)
    .plus(BigDecimal(picoRemainder))
    .longValueExact()

  /**
   * Provide the amount denominated in millisats (1,000th of a satoshi, 10 x pico btc).
   * Rounded towards zero if necessary.
   */
  fun millisat(): Long = BigDecimal(satoshi)
    .multiply(PICO_PER_SAT)
    .plus(BigDecimal(picoRemainder))
    .divide(PICO_PER_MSAT, RoundingMode.DOWN)
    .longValueExact()

  private val remainderAsBigDecimal: BigDecimal by lazy {
    BigDecimal(picoRemainder).divide(PICO_PER_BTC)
  }

  companion object {
    internal val MSAT_PER_SAT: BigDecimal = BigDecimal(1_000)
    internal val PICO_PER_MSAT: BigDecimal = BigDecimal(10)
    internal val PICO_PER_SAT: BigDecimal = BigDecimal(10_000)
    internal val SATS_PER_BTC: BigDecimal = BigDecimal(100_000_000)

    internal val PICO_PER_BTC: BigDecimal = SATS_PER_BTC.multiply(PICO_PER_SAT)

    internal const val SCALE_SATS = 8
    internal const val SCALE_PICO = 12

    // Due to Long overflow, this (1 million BTC) is the threshold at which we will never set a the scale above 8
    // when converting a BitcoinAmount to a CryptoAmount, even if a non-zero remainder_pico value is present e.g.
    // BitcoinAmount(sats = 1_000_000 * 100_000_000L, remainder_pico = 5555).toCryptoAmount() will strip the remainder.
    internal val SCALE_ROUNDING_LIMIT_IN_SATS = SATS_PER_BTC.multiply(BigDecimal(1_000_000))
  }
}

/** A BitcoinAmount representing the given millisat value (100,000,000,000 msat per btc) */
@Suppress("FunctionName")
fun MilliSatBitcoinAmount(milliSats: Long): BitcoinAmount {
  val input = BigDecimal(milliSats)
  val sats = input.divide(MSAT_PER_SAT).toLong()
  val pico = input.remainder(MSAT_PER_SAT).multiply(PICO_PER_MSAT).toInt()
  return BitcoinAmount(sats, pico)
}

/** A BitcoinAmount representing the given pico btc value (1,000,000,000,000 pico per btc) */
@Suppress("FunctionName")
fun PicoBitcoinAmount(picoBtc: Long): BitcoinAmount {
  val input = BigDecimal(picoBtc)
  val sats = input.divide(PICO_PER_SAT).toLong()
  val pico = input.remainder(PICO_PER_SAT).toInt()
  return BitcoinAmount(sats, pico)
}

/** Create a BitcoinAmount from the amount in BTC. Amounts below pico are truncated, not rounded. */
fun BigDecimal.toBitcoinAmount(): BitcoinAmount {
  val sats = this.multiply(SATS_PER_BTC).toLong()
  val pico = this.multiply(PICO_PER_BTC).remainder(PICO_PER_SAT).toInt()
  return BitcoinAmount(sats, pico)
}

internal val MSAT_PER_SAT: BigDecimal = BigDecimal(1_000)
internal val PICO_PER_MSAT: BigDecimal = BigDecimal(10)
internal val PICO_PER_SAT: BigDecimal = BigDecimal(10_000)
internal val SATS_PER_BTC: BigDecimal = BigDecimal(100_000_000)

internal val PICO_PER_BTC: BigDecimal = SATS_PER_BTC.multiply(PICO_PER_SAT)


lib/src/main/kotlin/app/cash/lninvoice/Bech32Data.kt
/*
 * Copyright (c) 2023 Block, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package app.cash.lninvoice

import app.cash.lninvoice.Bech32Data.Companion.Encoding.BECH32
import app.cash.lninvoice.Bech32Data.Companion.Encoding.BECH32M
import app.cash.quiver.extensions.ErrorOr
import app.cash.quiver.extensions.toEither
import arrow.core.Either
import arrow.core.flatMap
import arrow.core.left
import arrow.core.raise.either
import arrow.core.right
import okio.Buffer
import okio.ByteString
import okio.ByteString.Companion.toByteString
import java.util.Locale
import kotlin.streams.toList

data class Bech32Data(
  val encoding: Encoding,
  val hrp: String,
  val payload: ByteString
) {

  init {
    require(hrp.isNotEmpty()) { "hrp cannot be empty" }
    require(hrp.length <= 83) { "hrp is too long: ${hrp.length} > 83" }
  }

  private val checksumCanvas = byteArrayOf(0, 0, 0, 0, 0, 0)
  private val checksum: ByteString by lazy {
    val encoded = Buffer().also { buffer ->
      buffer.write(expand(hrp))
      buffer.write(payload)
      buffer.write(checksumCanvas)
    }.readByteString()

    val mod = polymod(encoded) xor encoding.const

    (0 until 6)
      .map { i -> ((mod shr (5 * (5 - i))) and 31).toByte() }
      .toByteArray()
      .toByteString()
  }

  val encoded: String by lazy {
    StringBuffer(hrp.length + payload.size + checksum.size + 1)
      .append(hrp.lowercase(Locale.ROOT))
      .append("1")
      .also { sb ->
        Buffer().apply {
          write(payload)
          write(checksum)
        }.readByteArray().forEach { b ->
          sb.append(encodingCharset[b.toInt()])
        }
      }.toString()
  }

  companion object {
    private const val encodingCharset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
    private val decodingCharset = mapOf(
      48 to 15, 50 to 10, 51 to 17, 52 to 21, 53 to 20, 54 to 26, 55 to 30, 56 to 7, 57 to 5, 65 to 29, 67 to 24,
      68 to 13, 69 to 25, 70 to 9, 71 to 8, 72 to 23, 74 to 18, 75 to 22, 76 to 31, 77 to 27, 78 to 19, 80 to 1,
      81 to 0, 82 to 3, 83 to 16, 84 to 11, 85 to 28, 86 to 12, 87 to 14, 88 to 6, 89 to 4, 90 to 2, 97 to 29,
      99 to 24, 100 to 13, 101 to 25, 102 to 9, 103 to 8, 104 to 23, 106 to 18, 107 to 22, 108 to 31, 109 to 27,
      110 to 19, 112 to 1, 113 to 0, 114 to 3, 115 to 16, 116 to 11, 117 to 28, 118 to 12, 119 to 14, 120 to 6,
      121 to 4, 122 to 2
    )

    private val parseRegex =
      Regex("^([!-@\\[-~]{1,83})1([ac-hj-np-z02-9]{6,})|([!-`{-~]{1,83})1([AC-HJ-NP-Z02-9]{6,})\$")

    private fun parse(encoded: String): ErrorOr<Pair<String, List<Int>>> =
      parseRegex.find(encoded)?.groupValues
        .toEither { IllegalArgumentException("Unparseable format") }
        .flatMap { matchResult ->
          val (hrp, data) =
            if (matchResult[1].isEmpty()) {
              matchResult[3].lowercase(Locale.ROOT) to matchResult[4]
            } else {
              matchResult[1] to matchResult[2]
            }
          Either.catch { data.chars().map(decodingCharset::getValue).toList() }
            .mapLeft { IllegalArgumentException("Invalid character") }
            .map { hrp to it }
        }

    fun decode(encoded: String): ErrorOr<Bech32Data> = either {
      val (hrp, data) = parse(encoded).bind()
      val values = data.map { it.toByte() }.toByteArray()
      val encoding = verifyChecksum(hrp, values).bind()
      Bech32Data(encoding, hrp, values.sliceArray(0 until values.size - 6).toByteString())
    }

    private fun verifyChecksum(
      hrp: String,
      values: ByteArray
    ): ErrorOr<Encoding> = when (
      val encoding = polymod(
        Buffer().also { buffer ->
          buffer.write(expand(hrp))
          buffer.write(values)
        }.readByteString()
      )
    ) {
      BECH32.const -> BECH32.right()
      BECH32M.const -> BECH32M.right()
      else -> InvalidChecksum(encoding).left()
    }

    private fun expand(hrp: String): ByteString =
      Buffer().also { buffer ->
        val ascii = hrp.chars().map { c -> c and 0x7f }.toList()
        ascii.forEach { buffer.writeByte((it shr 5) and 0x7f) }
        buffer.writeByte(0)
        ascii.forEach { buffer.writeByte(it and 0x1f) }
      }.readByteString()

    /** Find the polynomial with value coefficients mod the generator as 30-bit.  */
    private fun polymod(values: ByteString): Int {
      var c = 1
      values.toByteArray().forEach { vi ->
        val c0 = c ushr 25 and 0xff
        c = c and 0x1ffffff shl 5 xor (vi.toInt() and 0xff)
        if (c0 and 1 != 0) c = c xor 0x3b6a57b2
        if (c0 and 2 != 0) c = c xor 0x26508e6d
        if (c0 and 4 != 0) c = c xor 0x1ea119fa
        if (c0 and 8 != 0) c = c xor 0x3d4233dd
        if (c0 and 16 != 0) c = c xor 0x2a1462b3
      }
      return c
    }

    enum class Encoding(val const: Int) {
      BECH32(1),
      BECH32M(0x2bc830a3)
    }
  }
}

fun String.toBech32Data(): ErrorOr<Bech32Data> = Bech32Data.decode(this)

data class InvalidChecksum(val sum: Int) : Exception("Invalid checksum: $sum")
